proposals/000-decoupling-haddock-ghc.md

```diff
@@ -0,0 +1,137 @@
+
+Having an IR format whose parsing is de-coupled from GHC makes it more easily embeddable in other tools and libraries. This could lead, for example, to alternative (third-party) documentation-generation backends, where Haddock documentation can be rendered to e.g. Markdown, and thus more easily integrated into multi-language documentation systems. An IR format could also be leveraged to implement programs which transform documentation before rendering, just like [Pandoc filters](https://pandoc.org/filters).
+
+Moreover, having a well-defined IR format serializable to e.g. JSON, allows for tools written in other programming languages to interact with Haskell source code documentation. This might be important for tools such as [`haskell-spotlight`](https://github.com/haskell-spotlight/haskell-spotlight), a Visual Studio Code extension written in TypeScript.
+
+### Refactoring of Haddock
+
+We propose to refactor Haddock into GHC-specific and GHC-agnostic components. GHC-specific components include parsing of Haskell source files to find comments and interactions between GHC and Haddock. All other components are, in principle, GHC-agnostic:
+
+* Parsing comments into documentation;
+* Definition of the documentation syntax tree;
+* Querying and manipulating the documentation syntax tree;
+* Rendering of the documentation syntax tree to an output format such as HTML or LaTeX.
+
+Concretely, we propose:
+
+* Creating a package (tentatively named `haddock-backends`) dedicated to rendering backends. This package would contain the backend logic currently in `haddock-api`, refactored to remove GHC-specific datatypes.
+* Moving everything GHC-agnostic and unrelated to rendering backends to the `haddock-library` package.
+
+Factoring out GHC-specific components of Haddock will make it much easier to maintain. The GHC-agnostic components of Haddock will also be easier to integrate into other projects. Most importantly, if Haddock is restricted to take the IR format as an input, it will be possible to render documentation without needing GHC at all.
+
+## Timeline
+
+We expect the following tasks to be performed in order:
+
+1. Refactor Haddock such that rendering backends are decoupled from GHC internals and moved to a new package `haddock-backends`, GHC-specific components are kept in `haddock-api`, and all other GHC-agnostic components are moved to `haddock-library`;
+2. Design the IR file format, which is expressive enough to represent Haddock documentation and be backwards-compatible;
+3. Implement serialization/deserialization of the Haddock AST in `haddock-library`, allowing for embedding IR support into tooling;
+4. Update `haddock` executable to interact with the IR file format as both input and output.
+
+## People
+
+The people involved in executing this work are:
+
+* Laurent P. René de Cotret (@LaurentRDC)
+
+## Budget
+
+No funding from the Haskell Foundation is required to perform the work outlined in this proposal.
```

Comment:
If part 1 goes really well, I think this makes sense. However I fear part 1 will be quite tricky, and we'll have to think quite hard at what to do.

I don't want to saddle us with bureaucracy but I almost want two separate proposals. The first is an unbudgetted "Laurent et al. try to split out the backends, see what goes wrong". Then based on that information we can write a second proposal which quite likely is enough work that it will need funding and extra personnel.

We could do it with one proposal, but I suspect that will lead to a ton of errata. We could also just start step 1 right now and keep this PR open until we know more details, because doing phase 1 informally and just one proposal for phase 2.

Reply:
A 3rd option is to just keep this for "things go well, we at least want to try it" plan, but give it some sort of "effort box". Like if this proves to be quite hard, @LaurentRDC, it shouldn't be considered a failure to come back and say "I've spent all the free time I can, we need to do these X specific things to unblock Y, that requires more time". Merely uncovering that information is still successful!

Reply:
The errata process is only really there to correct severe factual errors in already-accepted proposals. Changes during execution are normal, expected, and don't need updates through this committee, unless the one doing the work feels like our feedback would be valuable. I think it's totally fine to work based on our best guess about the future here, and then if things end up differently, we'll have a conversation.